README.md
# flux-framework

Welcome to the web home of Flux Framework! Here we maintain:

- A central portal with information about active Flux Framework projects
- Automated pipelines to retrieve updated data for projects

## Usage

### Release Notes

The release notes in [_posts](_posts) are generated by the [rse-ops/releases-action](https://github.com/rse-ops/release-actions/tree/main/docs).
There are instructions there for how to use it, and if you want to run it locally.

### Data

We run a [nightly action](.github/workflows/update-data.yaml) to get updated data
for our projects, however you can generate the same data on your own using
[Contributor CI](https://contributor-ci.readthedocs.io/en/latest/getting_started/user-guide.html) as follows!

```bash
# Export a GitHub token to the environment
export GITHUB_TOKEN=xxxxxxxxxxx

# install contributor-ci
pip install contributor-ci

# check out the extractors available!
# cci list
```

The tool will read in repos from [contributor-ci.yaml](contributor-ci.yaml) (default) and we store data for cci in [_data](_data),
so to run an extractor you can do the following. Note that we want to cache data monthly so we use `--save-format year/month` instead of the
default of `year/month/day`

```bash
$ cci --out-dir _data extract --save-format year/month <extractor>
```

And the ones we do for the site (a subset) are (this is what we run in the action):

```bash
$ cci --out-dir _data extract --save-format year/month repos repo_metadata topics languages releases stars activity_commits activity_lines
```

Note that emojis and jekyll don't play nicely, so we currently don't have any in project descriptions.

### Api

The site data is also served from a set of static APIs! We do this in case you want data about Flux Framework
projects programmatically available, and as stated above, it's updated nightly.

 - `/flux-framework.github.io/api/languages.json`
 - `/flux-framework.github.io/api/releases.json`
 - `/flux-framework.github.io/api/repos.json`
 - `/flux-framework.github.io/api/repos-metadata.json`
 - `/flux-framework.github.io/api/stars.json`
 - `/flux-framework.github.io/api/topics.json`
 - `/flux-framework.github.io/api/commits.json`
 - `/flux-framework.github.io/api/lines.json`


### Custom Data

#### Avatars

If you want to define a custom avatar image for a project:

1. Add the image to [assets/img/custom/avatars](assets/img/custom/avatars)
2. Add _just the name_ to [_data/custom.yaml](_data/custom.yaml) under avatars.

#### Release

SPDX-License-Identifier: LGPL-3.0

LLNL-CODE-764420

iconv -c -f utf8 -t ascii  _data/data/latest/cci-repo_metadata.json > _data/data/latest/cci-repo_metadata.json



pages/releases.md
---
layout: page
title: Releases
permalink: /releases/
---

<input type="text" id="releaseSearch" onkeyup="searchReleases()" placeholder="Search for projects..">

  {% for post in site.posts %}
  <div class="post-preview">
    <h5 class="release-header" data-download="{{ post.url | prepend: site.baseurl }}" style="margin-bottom:4px; margin-top:4px;">
      <a href="{{ post.url | prepend: site.baseurl }}">{{ post.title }}</a> (<a href="{{ post.download }}" target="___blank">download</a>)
      <small style='margin-top:3px' class="pull-right">{{ post.date | date: '%B %d, %Y' }}</small>
    </h5>
  <div style="clear:both"></div>
  </div>
{% endfor %}

<script>
document.addEventListener('click', function (event) {

	// If the clicked element doesn't have the right selector, bail
	if (!event.target.matches('.release-header')) return;

	// Don't follow the link
	event.preventDefault();

	// Log the clicked element in the console
  url = event.target.getAttribute("data-download");
  document.location = url

}, false);

function searchReleases() {
  // Declare variables
  var input, filter, ul, li, a, i, txtValue;
  input = document.getElementById('releaseSearch');
  filter = input.value.toUpperCase();
  releases = document.getElementsByTagName('h5');

  // Loop through all list items, and hide those who don't match the search query
  for (i = 0; i < releases.length; i++) {
    a = releases[i].getElementsByTagName("a")[0];
    txtValue = a.textContent || a.innerText;
    if (txtValue.toUpperCase().indexOf(filter) > -1) {
      releases[i].style.display = "";
    } else {
      releases[i].style.display = "none";
    }
  }
}
</script>

src/gen-release-post.sh
#!/bin/sh
VERSION=$1
if test -z "$VERSION"; then
    echo "Usage: $0 VERSION [flux-core|flux-sched] [WORKDIR]" >&2
    exit 1
fi
NAME=${2:-flux-core}
WORKDIR=${3:-$(mktemp -d ${TMPDIR:-/tmp}/${NAME}.XXXXXX)}
SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"

if [[ -z "${WORKDIR}" ]]; then
    echo "Failed to make temporary work directory. Exiting"
    exit 1
fi
if [[ ! -d "${WORKDIR}/.git" ]]; then
    echo "Cloning $NAME $VERSION into $WORKDIR" >&2
    git clone --quiet --branch=v${VERSION} https://github.com/flux-framework/${NAME} ${WORKDIR}
    cd ${WORKDIR}
else
    cd ${WORKDIR}
    git checkout v${VERSION}
fi

URL="https://github.com/flux-framework/${NAME}/issues/"
TAG=$(git describe)
AUTHOR="flux-framework"
COMMIT=$(git rev-parse $TAG)
DATE=$( git cat-file -p $COMMIT \
      | gawk '/^tagger/{print strftime ("%F %H:%M:%S %z", $(NF-1))}')
DOWNLOAD="https://github.com/flux-framework/${NAME}/releases/tag/$TAG"

CALDATE=$( git cat-file -p $COMMIT \
      | gawk '/^tagger/{print strftime ("%F", $(NF-1))}')
OUTFILE="${SCRIPTDIR}/../_posts/${CALDATE}-$2-$VERSION.md"


# Front matter
cat > ${OUTFILE} <<EOF
---
layout: default
title: "${NAME} v$VERSION"
date: "$DATE"
author: "$AUTHOR"
categories: release
version: $VERSION
download: $DOWNLOAD
---

Download from GitHub [here]($DOWNLOAD)

# Release Notes
EOF
# Get top list of Release Notes for a single version
#   replace Issue #'s with links
#
awk "/^${NAME} version/{n++;getline;getline}; n<=1;" NEWS.md | \
 sed "s|#\([0-9][0-9]*\)|[#\1](${URL}\1)|g" >> ${OUTFILE}

if [[ -z "$3" ]]; then
    cd /tmp 
    rm -rf ${WORKDIR}
fi

assets/js/extractors/topics.js
---
layout: null
---

/* Creates word cloud visualization for webpage */
function draw_cloud_topics(areaID, repoNameWOwner, width, height) {

    width = width || stdTotalWidth
    height = height || stdTotalHeight

    // load data file, process data, and draw visualization
    var url = '{{ site.baseurl }}/api/topics.json';
    var files = [url];
    Promise.all(files.map(url => d3.json(url))).then(values => {
        var data = reformatData({"data": values[0]});
        drawCloud(data, areaID, width, height);
    });

    // Draw cloud from data
    function drawCloud(data, areaID, width, height) {
        var graphHeader = 'Flux Framework Topics';

        var wordScale = d3
            .scaleLinear()
            .domain([
                0,
                d3.max(data, function(d) {
                    return d.value;
                })
            ])
            .range([12, 60]);

        var fill = d3.scaleOrdinal(['#daa6d5', '#d69dd1', '#ac93cc', '#8988c7', '#7d7cc1', '#706fbb', '#4c4b63', '#4e5a84', '#5069a4', '#5386e4'])

        var margin = { top: stdMargin.top, right: stdMargin.right / 2, bottom: stdMargin.bottom / 2, left: stdMargin.left / 2 },
            width = stdTotalWidth - margin.left - margin.right,
            height = stdTotalHeight - margin.top - margin.bottom;

        var layout = d3.layout
            .cloud()
            .size([width, height])
            .words(
                data.map(function(d) {
                    return { text: d.name, size: wordScale(d.value) };
                })
            )
            .padding(5)
            .rotate(function() {
                return ~~(Math.random() * 2) * 90;
            })
            .fontSize(function(d) {
                return d.size;
            })
            .on('end', draw);

        layout.start();

        // Add title
        d3.select(areaID)
            .append('text')
            .attr('class', 'graphtitle')
            .attr('x', margin.left + width / 2)
            .attr('y', margin.top - margin.top / 3)
            .attr('text-anchor', 'middle')
            .text(graphHeader);

        function draw(words) {
            var wordCloud = d3
                .select(areaID)
                .attr('width', width + margin.left + margin.right)
                .attr('height', height + margin.top + margin.bottom)
                .append('g')
                .attr('width', width)
                .attr('height', height)
                .attr('transform', 'translate(' + (width / 2 + margin.left) + ',' + (height / 2 + margin.top) + ')')
                .selectAll('text')
                .data(words)
                .enter()
                .append('text')
                .attr('class', 'cloudtext')
                .style('font-size', function(d) {
                    return d.size + 'px';
                })
                .style('fill', function(d, i) {
                    return fill(d.size);
                })
                .attr('text-anchor', 'middle')
                .attr('transform', function(d) {
                    return 'translate(' + [d.x, d.y] + ')rotate(' + d.rotate + ')';
                })
                .text(function(d) {
                    return d.text;
                });
        }
    }

    // Turn json obj into desired word list
    function reformatData(obj) {
        var wordDict = {};
        var repos = repoNameWOwner == null ? Object.keys(obj['data']) : [repoNameWOwner];
        repos.forEach(function(repo) {
            if (obj['data'].hasOwnProperty(repo)) {
                var topicNodes = obj['data'][repo]['repositoryTopics']['nodes'];
                for (var i = 0; i < topicNodes.length; i++) {
                    var aWord = topicNodes[i]['topic']['name'];
                    if (!Object.keys(wordDict).contains(aWord)) {
                        wordDict[aWord] = 0;
                    }
                    wordDict[aWord] += 1;
                }
            }
        });
        var data = [];
        for (var aWord in wordDict) {
            if (wordDict.hasOwnProperty(aWord)) {
                var datpair = { name: aWord, value: wordDict[aWord] };
                data.push(datpair);
            }
        }
        // Prioritize highest counts
        data.sort((a, b) => (a.value < b.value ? 1 : a.value > b.value ? -1 : 0));
        return data;
    }
}


assets/js/extractors/languages.js
---
layout: null
---

/* Creates pie chart visualization for webpage */
function draw_pie_languages(areaID, width, height) {

    width = width || stdTotalWidth
    height = height || stdTotalHeight

    var url = '{{ site.baseurl }}/api//languages.json';
    var files = [url];
    Promise.all(files.map(url => d3.json(url))).then(response => {
            var data = reformatData({"data": response[0]});
            drawGraph(data, areaID, width, height);
        });

    // Draw graph from data
    function drawGraph(data, areaID, width, height) {
        var graphHeader ='Flux Framework Languages';

        var margin = { top: 8, right: 8, bottom: 8, left: 8 },
            width = width - margin.left - margin.right,
            height = height - margin.top - margin.bottom,
            radius = d3.min([width - margin.left - margin.right, height - margin.top - margin.bottom]) / 2,
            donutWidth = 100;

        var tip = d3
            .tip()
            .attr('class', 'd3-tip')
            .offset(function() {
                return [-10, 0];
            })
            .html(function(d) {
                var units = ' MB';
                var divisor = 1000000;
                if (d.data.size > 1000000000) {
                    divisor *= 1000;
                    units = ' GB';
                }
                return d3.format('.2f')(d.data.size / divisor) + units + '<br>' + d.data.name;
            });

        var chart = d3
            .select(areaID)
            .attr('width', width)
            .attr('height', height)
            .append('g')
            .attr('transform', 'translate(' + (width / 2 + margin.left) + ',' + (height / 2 - margin.top) + ')');

        chart.call(tip);

        var color = d3.scaleLinear([0, 2 * Math.PI], ['#036291', '#94c5dc']);

        var arc = d3
            .arc()
            .innerRadius(radius - donutWidth)
            .outerRadius(radius)
            .padAngle(0.002);

        var pie = d3
            .pie()
            .value(function(d) {
                return d.size;
            })
            .sort(null);

        const pathGroup = chart.append('g');
    
        var path = pathGroup
            .selectAll('path')
            .data(pie(data[2]))
            .join('path')
            .attr('d', arc)
            .attr('fill', d => color(d.startAngle))
            .style('cursor', 'pointer')
            .on('mouseover', tip.show)
            .on('mouseout', tip.hide)
            .on('click', clicked);

        function clicked(d) {
            const pieData = pie(data[d.data.reference.number][d.data.reference.key]);
            const scaleCoeff = (d.endAngle - d.startAngle)/(2 * Math.PI);
            let shift = d.startAngle;
            let delta = 0;
            pieData.forEach(o => {
                delta = (o.endAngle - o.startAngle) * scaleCoeff;
                o.current = { startAngle: shift, endAngle: shift + delta }
                o.target = { startAngle: o.startAngle, endAngle: o.endAngle }
                shift += delta;
            });

            path = pathGroup
                .selectAll('path')
                .data(pieData)
                .join('path')
                .attr('d', arc)
                .attr('fill', color(d.startAngle))
                .style('cursor', 'pointer')
                .on('mouseover', tip.show)
                .on('mouseout', tip.hide)
                .on('click', clicked);
            
            label
                .selectAll('text')
                    .data(pieData)
                    .join('text')
                        .attr('dy', '0.35em')
                        .attr('font-size', '11px')
                        .attr('fill-opacity', 0)
                        .text(o => o.data.name);

            const dur = 1000;

            const t = chart.transition().duration(dur);

            path.transition(t)
                .tween('data', d => {
                    const i = d3.interpolate(d.current, d.target);
                    return t => d.current = i(t);
                })
                .attr('fill', d => color(d.startAngle))
                .attrTween("d", d => () => arc(d.current));

            d3.select('#mainTitle').attr('font-size', '18px').text(d.data.name + ' breakdown');

            label.selectAll('text').filter(d => labelVisible(d.target)).transition(t)
                .attr('fill-opacity', 1)
                .attrTween('transform', d => () => labelTransform(d.current));

            labelDynamicSizing();
            titleResizing();
        }

        function middleClicked() {
            path = pathGroup
                .selectAll('path')
                .data(pie(data[2]))
                .join('path')
                .attr('d', arc)
                .attr('fill', d => color(d.startAngle))
                .style('cursor', 'pointer')
                .on('mouseover', tip.show)
                .on('mouseout', tip.hide)
                .on('click', clicked);

            label
            .selectAll('text')
                .data(pie(data[2]))
                .join('text')
                    .attr('dy', '0.35em')
                    .attr('font-size', '11px')
                    .attr('fill-opacity', 0)
                    .text(o => o.data.name);

            label.selectAll('text').filter(d => labelVisible(d))
                .attr('fill-opacity', 1)
                .attr('transform', d => labelTransform(d));

            d3.select('#mainTitle').attr('font-size', '18px').text(graphHeader);
        }

        const textArray = [{ text: graphHeader, size: 18, weight: 'bold', id: 'mainTitle', x: 0, y: 0 }, { text: 'by bytes', size: 18, weight: 'normal', id: 'subTitle', x: 0, y: '1em' }];
        
        // Add title
        const titles = chart
            .append('g')
            .selectAll('text')
                .data(textArray)
                .join('text')
                    .attr('font-size', d => d.size + 'px')
                    .attr('font-weight', d => d.weight)
                    .attr('id', d => d.id)
                    .attr('x', d => d.x)
                    .attr('y', d => d.y)
                    .attr('text-anchor', 'middle')
                    .text(d => d.text)
                    .on('click', middleClicked);

        // Adds labels to wedges
        const label = chart
            .append('g')
                .style('font-size', '11px')
                .attr('pointer-events', 'none')
                .attr('text-anchor', 'middle')
                .style('user-select', 'none');
        
        label
            .selectAll('text')
                .data(pie(data[2]))
                .join('text')
                    .attr('dy', '0.35em')
                    .attr('fill-opacity', 0)
                    .text(o => o.data.name);

        label.selectAll('text').filter(d => labelVisible(d))
            .attr('fill-opacity', 1)
            .attr('transform', d => labelTransform(d));

        const center = chart
            .append('circle')
            .attr('r', radius - donutWidth)
            .attr('fill', 'none')
            .attr('pointer-events', 'all')
            .style('cursor', 'pointer')
            .on('click', middleClicked);

        labelDynamicSizing();

        function labelVisible(d) {
            return (d.endAngle - d.startAngle) > 0.07;
        }
    
        function labelTransform(d) {
            const x = (d.endAngle + d.startAngle) / 2 * 180 / Math.PI;
            const y = radius - donutWidth / 2;
            return `rotate(${x - 90}) translate(${y},0) rotate(${x < 180 ? 0 : 180})`;
        }

        function labelDynamicSizing() {
            label.selectAll('text').nodes().forEach(label => {
                if (label.getComputedTextLength() > donutWidth) {
                    label.setAttribute('font-size', 11 * donutWidth / (label.getComputedTextLength() + 4) + 'px');
                }
            })
        }

        function titleResizing() {
            titles.nodes().forEach(label => {
                if (label.getComputedTextLength() > 2 * (radius - donutWidth)) {
                    label.setAttribute('font-size', 36 * (radius - donutWidth) / (label.getComputedTextLength() + 4) + 'px');
                }
            })
        }
    }

    // Turn json obj into desired working data
    function reformatData(obj) {
        var languageData = {};
        var repoData = {};

        // Process repo breakdown by language
        Object.keys(obj['data']).forEach(repoName => {
            repoData[repoName] = [];
            var value = obj['data'][repoName]['languages']['edges'] != null ? obj['data'][repoName]['languages']['edges'].length : 0;
            for (var i = 0; i < value; i++) {
                repoData[repoName].push({ name: obj['data'][repoName]['languages']['nodes'][i]['name'], size: obj['data'][repoName]['languages']['edges'][i]['size'], reference: { number: 1, key: obj['data'][repoName]['languages']['nodes'][i]['name'] } });
            }
            repoData[repoName].sort((a, b) => b.size - a.size);
        });

        // Process language breakdown
        Object.keys(repoData).forEach(key => {
            repoData[key].forEach(language => {
                if (languageData.hasOwnProperty(language.name)) {
                    languageData[language.name].push({ name: key, size: language.size, reference: { number: 0, key: key } });
                } else {
                    languageData[language.name] = [{ name: key, size: language.size, reference: { number: 0, key: key } }];
                }
            });
        });
        Object.keys(languageData).forEach(key => {
            languageData[key].sort((a, b) => b.size - a.size);
        });

        // Process total breakdown
        var totalData = [];
        Object.keys(languageData).forEach(key => {
            totalData.push({ name: key, size: sum(languageData[key]), reference: { number: 1, key: key } });
        });
        totalData.sort((a, b) => b.size - a.size);

        return [repoData, languageData, totalData];
    }

    function sum(inputArray) {
        var total = 0;
        inputArray.forEach(d => {
            total += d.size;
        });
        return total;
    }
}


assets/js/extractors/licenses.js
---
layout: null
---

/* Creates a zoomable sunburst visualization for webpage */
function draw_sunburst_licenses(areaID, width, height) {

    width = width || stdTotalWidth
    height = height || stdTotalHeight

    var url = '{{ site.baseurl }}/api/repos.json';
    var files = [url];
    Promise.all(files.map(url => d3.json(url))).then(values => drawSunburst({"data": values[0]}, areaID, width, height));

    function drawSunburst(obj, areaID, setwidth, setheight) {
        let data = reformatData(obj, null);

        const graphHeader = 'Flux Framework Licenses';

        const margin = { top: stdMargin.top, right: stdMargin.right, bottom: stdMargin.bottom / 2, left: stdMargin.left },
            width = setwidth * 2 - margin.left - margin.right,
            height = setheight * 2 - margin.top - margin.bottom;

        const radius = Math.min(width, height) / 6;
        const partition = data =>  {
            const root = d3
            .hierarchy(data)
            .sum(d => d.value)
            .sort((a, b) => b.value - a.value);

            return d3.partition()
                .size([2 * Math.PI, root.height + 1])
            (root);
        }

        let root = partition(data);

        const tip = d3
            .tip()
            .attr('class', 'd3-tip')
            .offset([-10, 0])
            .html(function(d) {
                return `${d.data.name}`;
            });

        const arc = d3
            .arc()
                .startAngle(d => d.x0)
                .endAngle(d => d.x1)
                .padAngle(d => Math.min((d.x1 - d.x0) / 2, 0.005))
                .padRadius(radius * 1.5)
                .innerRadius(d => d.y0 * radius)
                .outerRadius(d => Math.max(d.y0 * radius, d.y1 * radius));

        const colors = d3.scaleOrdinal()
            .domain([0, root.children.length - 1])
            .range(["#8dd3c7","#ffffb3","#bebada","#fb8072","#80b1d3","#fdb462","#b3de69","#fccde5","#d9d9d9","#bc80bd","#ccebc5","#ffed6f"]);
        
        const chart = d3
            .select(areaID)
                .attr('width', width + margin.left + margin.right)
                .attr('height', height + margin.top + margin.bottom)

        chart.call(tip);

        // Adds title
        chart
            .append('g')
                .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')')
                .append('text')
                    .attr('class', 'graphtitle')
                    .attr('x', width / 2)
                    .attr('y', 0 - margin.top / 3)
                    .attr('text-anchor', 'middle')
                    .text(graphHeader);

        // Adds labels in center
        const licenseNames = chart
            .append('g')
                .attr('id', 'licenseNames')
                .attr('transform', `translate(${margin.left + width / 2},${margin.top + height / 2})`)
                .attr('pointer-events', 'none')
                .selectAll('text')
                    .data(root.children)
                    .join('text')
                        .attr('text-anchor', 'middle')
                        .attr('font-size', '14px')
                        .attr('fill-opacity', 0)
                        .attr('dy', '0.35em')
                        .attr('id', d => `licenseName${d.data.name}`)
                        .text(d => d.data.name);

        root.each(d => d.current = d);

        licenseNames.attr('fill-opacity', 0);

        const centerGroup = chart
            .append('g')
                .attr('transform', () => {
                    let xcoord = width / 2 + margin.left;
                    let ycoord = height / 2 + margin.top;
                    return 'translate(' + xcoord + ',' + ycoord + ')';
                })
                .attr('id', 'licenseCenter');

        // Creates wedges of sunburst
        const path = centerGroup
            .append('g')
            .selectAll('path')
                .data(root.descendants().slice(1))
                .join('path')
                    .attr('fill', d => {
                        while (d.depth > 1) {
                            d = d.parent;
                        }
                        return colors(d.parent.children.indexOf(d));
                    })
                    .attr('fill-opacity', 0.9)
                    .attr('d', d => arc(d.current))
                    .on('mouseover', tip.show)
                    .on('mouseout', tip.hide);
        
        // Makes license wedges appear clickable
        path.filter(d => d.children)
            .style('cursor', 'pointer')
            .on('click', clicked);

        path.filter(d => !d.children)
            .style('cursor', 'pointer')
            .on('click', d => {
                const win = window.open(`${window.location['origin']}/repo/#/${d.data.name.split('/')[0]}/${d.data.name.split('/')[1]}`);
                win.focus();
            });

        // Adds labels to wedges
        let label = centerGroup
            .append('g')
                .style('font-size', '11px')
                .attr('pointer-events', 'none')
                .attr('text-anchor', 'middle')
                .style('user-select', 'none')
                .selectAll('text')
                    .data(root.descendants().slice(1))
                    .join('text')
                        .attr('dy', '0.35em')
                        .attr('fill-opacity', d => +labelVisible(d.current))
                        .attr('transform', d => labelTransform(d.current))
                        .text(d => {
                            let name = d.data.name;
                            if (d.current.height > 0) {
                                return name;
                            } else {
                                name = name.split('/')[1];
                                if(name.length > 20) {
                                    return '...';
                                } else {
                                    return name;
                                }
                            }
                        });
        
        // Creates blank circle in center with click event to go up one level
        const parent = centerGroup
            .append('circle')
            .datum(root)
            .attr('r', radius)
            .attr('fill', 'none')
            .attr('pointer-events', 'all')
            .on('click', clicked);

        // Makes center label fully visible
        d3.select('#licenseNames').raise();

        let lastClickedObject = root;

        function clicked(o) {
            lastClickedObject = o;

            let newLabel = label.filter(d => labelEverVisible(d));

            parent.datum(o.parent || root)

            root.each(d => d.target = {
                x0: Math.max(0, Math.min(1, (d.x0 - o.x0) / (o.x1 - o.x0))) * 2 * Math.PI,
                x1: Math.max(0, Math.min(1, (d.x1 - o.x0) / (o.x1 - o.x0))) * 2 * Math.PI,
                y0: Math.max(0, d.y0 - o.depth),
                y1: Math.max(0, d.y1 - o.depth)
            });

            const dur = 1000;

            const t = centerGroup.transition().duration(dur);

            path.transition(t)
                .tween('data', d => {
                    const i = d3.interpolate(d.current, d.target);
                    return t => d.current = i(t);
                })
                .attrTween("d", d => () => arc(d.current));

            newLabel.transition(t)
                .attr('fill-opacity', d => +labelVisible(d.target))
                .attrTween('transform', d => () => labelTransform(d.current));

            licenseNames.transition()
                .duration(dur)
                .attr('fill-opacity', d => +(d.data.name == o.data.name));

        }

        function labelVisible(d) {
            return d.y1 <= 3 && d.y0 >= 1 && (d.x1 - d.x0) > 0.07;
        }

        function labelEverVisible(d) {
            const scale = 2 * Math.PI / (d.parent.x1 - d.parent.x0);
            return (d.x1 - d.x0) * scale > 0.07;
        }
    
        function labelTransform(d) {
            const x = (d.x0 + d.x1) / 2 * 180 / Math.PI;
            const y = (d.y0 + d.y1) / 2 * radius;
            return `rotate(${x - 90}) translate(${y},0) rotate(${x < 180 ? 0 : 180})`;
        }

        let options = ['# of Stars per Repo', '# of Forks per Repo', '# of Contributors per Repo', '# of Repos'];

        const slider = d3
            .sliderRight()
            .domain([0,3])
            .step(1)
            .tickFormat(d => {
                return options[Math.round(d)];
            })
            .ticks(3)
            .on('onchange', val => {
                data = reformatData(obj, val);
                root = partition(data);
                root.each(d => d.current = d);
                path.data(root.descendants().slice(1)).join(enter => enter, update => {
                    update.attr('fill', d => {
                                while (d.depth > 1) {
                                    d = d.parent;
                                }
                                return colors(d.parent.children.indexOf(d));
                            })
                            .attr('fill-opacity', 0.9)
                            .attr('d', d => {
                                if (lastClickedObject.depth == 0) {
                                    return arc(d.current);
                                } else {
                                    let o = root.children.filter(d => d.data.name == lastClickedObject.data.name)[0];
                                    root.each(d => d.current = {
                                        x0: Math.max(0, Math.min(1, (d.x0 - o.x0) / (o.x1 - o.x0))) * 2 * Math.PI,
                                        x1: Math.max(0, Math.min(1, (d.x1 - o.x0) / (o.x1 - o.x0))) * 2 * Math.PI,
                                        y0: Math.max(0, d.y0 - o.depth),
                                        y1: Math.max(0, d.y1 - o.depth)
                                    });
                                    return arc(d.current);
                                }
                            });
                    return update;
                }, exit => exit);
                label.data(root.descendants().slice(1))
                    .join('text')
                        .attr('dy', '0.35em')
                        .attr('fill-opacity', d => +labelVisible(d.current))
                        .attr('transform', d => labelTransform(d.current))
                        .text(d => {
                            let name = d.data.name;
                            if (d.height > 0) {
                                return name;
                            } else {
                                name = name.split('/')[1];
                                if(name.length > 20) {
                                    return '...';
                                } else {
                                    return name;
                                }
                            }
                        });
                path.selectAll('title').remove();
                path.append('title').text(d => d.data.name);
            });
    
        // Creates option slider
        chart.append('g')
            .attr('transform', `translate(${margin.left},${margin.top / 2})`)
            .attr('id', 'licenseSlider')
            .call(slider);

    }

    // Turns json into useable data. Output MUST use arrays in accordance with the d3.hierarchy requirements
    function reformatData(obj, type) {
        // Converts json file to object with necessary information for later processing
        const objTree = {};
        const keys = Object.keys(obj['data']);
        for (var repo of keys) {
            let licenseInfo = obj['data'][repo]['licenseInfo'];
            // Skip if no license info
            if (licenseInfo == null) {
                continue;
            }
            // Use long name if no short name
            let licenseName = licenseInfo['spdxId'] == null || licenseInfo['spdxId'] == 'NOASSERTION' ? licenseInfo['name'] : licenseInfo['spdxId'];
            // Skip if no license name
            if (licenseName == null) {
                continue;
            }
            if (objTree.hasOwnProperty(licenseName)) {
                objTree[licenseName][repo] = repo;
            } else {
                objTree[licenseName] = {};
                objTree[licenseName][repo] = repo;
            }
        }

        // Converts object tree into d3.hierarchy compliant form
        const licenseArray = [];
        const licenses = Object.keys(objTree);
        licenses.forEach(function(license) {
            const repoArray = [];
            const repos = Object.keys(objTree[license]);
            repos.forEach(function(repo) {
                // Chooses what value to use for weights
                let value = 1;
                if (type == null || type == 0) {
                    value = obj['data'][repo]['stargazers']['totalCount'];
                } else if (type == 1) {
                    value = obj['data'][repo]['forks']['totalCount'];
                } else if (type == 2) {
                    value = obj['data'][repo]['mentionableUsers']['totalCount'];
                }
                repoArray.push({ name: repo, value: value });
            })
            licenseArray.push({ name: license, children: repoArray });
        });
        const data = { name: 'Licenses', children: licenseArray };
        return data;
    }
}
